### DISCUSSION TITLE
Изменение Api селектов

### DISCUSSION THREAD

**Discussion by gizeasy** [consta-design-system/uikit#1687] [Ideas]
# 1. Переименовать `id` в `value` (сохранив обратную совместимость, временно поддерживая поле id).

### Сейчас:
```tsx
type Item = {
  id: number;
  label: string;
};

const items: Item[] = [
  {
    id: 1,
    label: '1',
  },
  {
    id: 2,
    label: '2',
  },
  {
    id: 3,
    label: '3',
  },
];

const App = () => {
    const [value, setValue] = useState<Item | null | undefined>();
    return <Select items={items} value={value} onChange={({ value }) => setValue(value)} />;
}
```

### Будет:
```tsx
type Item = {
  id: number;
  label: string;
};

const items: Item[] = [
  {
    value: 1,
    label: '1',
  },
  {
    id: 2,
    label: '2',
  },
  {
    id: 3,
    label: '3',
  },
];

const App = () => {
    const [value, setValue] = useState<Item | null | undefined>();
    return <Select items={items} value={value} onChange={({ value }) => setValue(value)} />;
}
```

или возможно будет писать и так и так

# 2. Изменить тип value (в прошлом id) на string или оставить как сейчас - generic тип.

### Сейчас:
Есть возможность указать как `strung` так `number` в `value`  item'а:
```tsx
type Item = {
  value: number;
  label: string;
};

const items: Item[] = [
  {
    value: 1,
    label: '1',
  },
  {
    value: 2,
    label: '2',
  },
  {
    value: 3,
    label: '3',
  },
];

const App = () => {
    const [value, setValue] = useState<Item | null | undefined>();
    return <Select items={items} value={value} onChange={({ value }) => setValue(value)} />;
}
```

### Будет:
Не будет возможности указать в `value` item'а тип `number` возможно будет использовать только строки.

# 3. Что передаем в поле `value` (строка, целиком item либо и то, и другое)

### Сейчас:
В значение инпута можно указать только целиком item:
```tsx
type Item = {
  value: string;
  label: string;
};

const items: Item[] = [
  {
    value: '1',
    label: '1',
  },
  {
    value: '1',
    label: '2',
  },
  {
    value: '1',
    label: '3',
  },
];

const App = () => {
    const [value, setValue] = useState<Item | null | undefined>();
    return <Select items={items} value={value} onChange={({ value }) => setValue(value)} />;
}
```

### Будет:
В значение инпута можно будет указать `value` item'а:
```tsx
type Item = {
  value: string;
  label: string;
};

const items: Item[] = [
  {
    value: '1',
    label: '1',
  },
  {
    value: '1',
    label: '2',
  },
  {
    value: '1',
    label: '3',
  },
];

const App = () => {
    const [value, setValue] = useState<string | null | undefined>();
    return <Select items={items} value={value} onChange={({ value }) => setValue(value.value)} />;
}
```

Или оставить и так и так

# 4. Объект, который приходит в onChange (либо возвращаем value, либо item, либо объект event c полем data)

### Сейчас:
```tsx
type onChangeProps<ITEM> = {
   e: React.SyntheticEvent;
   value: ITEM;
}
```

### Предложение 1
```tsx
type onChangeProps<ITEM> = {
   e: React.SyntheticEvent;
   value: string;
   item: ITEM;
}
```

### Предложение 2
```tsx
type onChangeProps = {
   e: React.SyntheticEvent;
   value: string;
}
```

### Предложение 3

возвращать только event, у которого в поле `data` будет значение выбранного поля `string`

**Comment by DmitriyWebDev** [comment]
**По предложению №1.** 
**Переименовать id в value (сохранив обратную совместимость, временно поддерживая поле id).** 

Поддерживаю, но думаю, что не стоит затягивать с поддержкой обратной совместимости поля id, чтобы был единый контракт для работы с селектами.

**По предложению №2.** 
**Изменить тип value (в прошлом id) на string или оставить как сейчас - generic тип.**

Считаю, что нужно оставить generic тип, так как это позволяет гибко работать со значениями, цифры в любой момент могут понадобиться в value.

**По предложению №3.** 
**Что передаем в поле value (строка, целиком item либо и то, и другое)**

Думаю, что будет правильнее оставить как есть сейчас, то есть в значение селекта указывать только целиком item и отображать выбранный item(s) по его value(s). 
Таким путем пошел [react-select ](https://github.com/JedWatson/react-select) и пользоваться этим удобно.
В коде нужно просто передавать item целиком, не вникая в его поля.

**По предложению №4.** 
**Объект, который приходит в onChange (либо возвращаем value, либо item, либо объект event c полем data)**

Думаю, что наилучший вариант уже реализован сейчас и менять ничего не нужно.
Так как для любых ситуаций получаемой информации достаточно.

**Reply by VladislavYukhimenko** [reply]
По предложению №3.
Если мы в value будем хранить option, то для восстановления выбранного value из данных нам нужно будет обходить все options и искать тот, у которого value соответствует нашему value.
Также стоит обратить внимание на то, что реакт оперирует иммутабельными данными, а это значит, что мы можем получить рассинхрон в том option, который хранится в стейте с теми, что лежит в списке options. 
Например:
1. мы сохранили в state option как value
2. система обновила список options, добавив поле disabled к тому значению, которое мы сохранили на первом шаге (при этом создается новый объект)
3. Селект не может найти value, т.к. в нашем стейте все еще лежит старый объект, а в списке опшнов - обновленный путем пересоздания (ссылки не совпадают)

**Reply by DmitriyWebDev** [reply]
По предложению №3.

Там как раз нужно будет искать выбранный элемент(ы) селекта не по ссылке, а по значению его (их) ключевого поля - Item.value, а это уже примитив (строка или число).

Поэтому проблема рассинхрона тут не возникнет, так как проверка на выбранный элемент реализована именно через проверку примитивов - Item.value.

При добавлении в выбранный элемент селекта нового поля "disabled" также ничего не сломается и не изменится в логике поиска выбранного элемента(ов).

**Reply by VladislavYukhimenko** [reply]
это ведь не явное поведение. т.е. мы в value передаем 1 объект, а компонент выбирает другой, основываясь на сравнении ключевого поля

**Reply by monolithed** [reply]
Семантически в value должно передаваться только то, что будет отражено, т.е. строка. Смысл передавать туда что-то другое?

**Reply by DmitriyWebDev** [reply]
> это ведь не явное поведение. т.е. мы в value передаем 1 объект, а компонент выбирает другой, основываясь на сравнении ключевого поля

Сравнение по value, на мой взгляд, это очевидное и явное поведение, так как это ключевое поле-примитив.
Тут все прозрачно.

**Comment by VladislavYukhimenko** [comment]
я за:
1. Переименовать id в value
2. сделать value - generic
3. в onChange - получать именно value, а не option целиком (+ возможно ивент)
4. в проперти value прокидывать именно value, а не option целиком

+ было бы не плохо убрать 100500 пропсов, позволяющих кастомизировать поведение селекта. Чем инструмент проще - тем меньше проблем

**Reply by gizeasy** [reply]
2. Дженерик ограничить string | number?

**Reply by VladislavYukhimenko** [reply]
нет, дженерик может быть любым объектом

**Reply by monolithed** [reply]
Дженерики это прикольно, но по спецификации value возвращает строку, что также решает проблему с проверкой значения — не нужно писать ифаки и пр. лабуду, потому что обращение даже к пустому value всегда вернет строку.

**Reply by VladislavYukhimenko** [reply]
value вполне может быть объектом, иногда это бывает удобно:
```
const options = [
  {
    label: 'Peter',
    value: {
      name: 'Peter',
      lastName: 'Vasiliev',
      age: 20,
    },
  },
  {
    label: 'Ivan',
    value: {
      name: 'Ivan',
      lastName: 'Ivanov',
      age: 22,
    },
  },
  {
    label: 'Natasha',
    value: {
      name: 'Natasha',
      lastName: 'Romanovich',
      age: 18,
    },
  },
];

const PersonSelector = () => {
  const [person, setPerson] = useState(options[0].value);

  /** В value лежит именно option.value, а не option целиком. Т.е.:
   *   {
   *     name: 'Peter',
   *     lastName: 'Vasiliev',
   *     age: 20,
   *   },
   */
  return <Select options={options} value={person} onChange={setPerson} />;
};

```
Не совсем понятно в каком кейсе появляются ифы

**Comment by monolithed** [comment]
По п. 4. обратите внимание, что передача Event с полем detail — это [стандартный способ](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/detail), который позволяет:

1. передать данные общеизвестным способом модификации типов
2. хранить типы событий в общеизвестном месте Event.type

**Comment by aleksandr-potkin-nedra** [comment]
П.4 Для интеграции со сторонними библиотеками форм удобно использовать такой тип onChange:
type OnChangeType<T> = (e: { target: { name: string; value: T } }, value: T) => void;

**Reply by monolithed** [reply]
`target` обычно указывает на DOM-объект (что может быть полезно), а для передачи данных есть стандартное свойство `detail`, которое находится в `UIEvent.detail` и `CustomEvent.detail`. `name` это тип события? Если да, то чем не угодил `type`?

**Reply by monolithed** [reply]
Что может быть проще стандартной формы работы с событиями?

```ts
type ChangeEvent<T> = React.SyntheticEvent & {
     type: '....',
     detail: {
           value: T
     }
};

const onChange = (event: ChangeEvent<T>): void => { ... }
```

**Reply by aleksandr-potkin-nedra** [reply]
name - уникальное имя элемента формы или поля в форме куда записывается значение

**Reply by VladislavYukhimenko** [reply]
кажется, что для интеграции со сторонними библиотеками можно написать свой хелпер на onChange и не добавлять в селект лишнюю логику:

```
const handleWithForm = (key: string) => (value) => {
 changeFormState({
  name: key,
  value,
 })
}
```

Тогда использование селекта будет выглядеть следующим образом:

```
<Select value={formData.key} onChange={handleWithForm('key')} />
```
